src/log.rs

```diff
@@ -162,4 +165,76 @@ impl Writer {
 
         Ok(())
     }
+
+    pub fn compact_log<S: Serialize, P: AsRef<Path>>(
+        &self,
+        path: P,
+        data: Vec<S>,
+    ) -> Result<(), Error> {
+        let new_db_path = path.as_ref().with_file_name(Uuid::new_v4().to_string());
+
+        let mut new_db = OpenOptions::new()
+            .read(true)
+            .create(true)
+            .append(true)
+            .open(&new_db_path)
+            .map_err(|err| {
+                Error::OsError(
+                    format!(
+                        "While creating compaction file at {:?}, we received an error from the OS: {}",
+                        new_db_path.as_ref(),
+                        err
+                    ),
+                    err,
+                )
+            })?;
+
+        let mut data = bincode::serialize(&LogItems::Batch(data))
+            .map_err(|err| Error::serialize(std::any::type_name::<LogItems<S>>(), err))?;
+
+        let size = data.len() as u32;
+
+        let mut to_write = size.to_be_bytes().to_vec();
+        to_write.append(&mut data);
+
+        new_db.write_all(&to_write).map_err(|err| {
+            Error::OsError(
+                format!(
+                    "Failed to append {} bytes to the log for log compaction, error: {}",
+                    to_write.len(),
+                    err
+                ),
+                err,
+            )
+        })?;
+
+        let old_db_path = path.as_ref().with_file_name(Uuid::new_v4().to_string());
+
+        fs::rename(&path, &old_db_path).map_err(|err| {
+            Error::OsError(
+                format!(
+                    "Failed to rename the old schema file to a temporary name: {:?}, error: {:?}.",
+                    old_db_path.to_str(),
+                    err
+                ),
+                err,
+            )
+        })?;
+        fs::rename(&new_db_path, &path).map_err(|err| Error::OsError(format!("Failed to rename the new schema file to the appropriate schema name: {:?}, error: {:?}.", path.as_ref().to_str(), err), err))?;
+        fs::remove_file(old_db_path).map_err(|err| {
+            Error::OsError(
+                format!("Failed to clean up old schema file, error: {:?}.", err),
+                err,
+            )
+        })?;
```

Comment:
Why are you doing `mv`, `mv`, `rm`? You can just do a single `rename`. You also don't need to recalculate `old_db_path` you already know the value of this.